cmd/sops/main.go

```diff
@@ -557,77 +564,115 @@ func edit(c *cli.Context, file string, fileBytes []byte) ([]byte, error) {
 	}
-	tmpdir, err := ioutil.TempDir("", "")
-	if err != nil {
-		return nil, cli.NewExitError(fmt.Sprintf("Could not create temporary directory: %s", err), exitCouldNotWriteOutputFile)
-	}
-	defer os.RemoveAll(tmpdir)
-	tmpfile, err := os.Create(path.Join(tmpdir, path.Base(file)))
-	if err != nil {
-		return nil, cli.NewExitError(fmt.Sprintf("Could not create temporary file: %s", err), exitCouldNotWriteOutputFile)
-	}
-	var out []byte
-	if c.Bool("show-master-keys") {
-		out, err = outputStore(c, file).MarshalWithMetadata(tree.Branch, tree.Metadata)
-	} else {
-		out, err = outputStore(c, file).Marshal(tree.Branch)
-	}
-	if err != nil {
-		return nil, cli.NewExitError(fmt.Sprintf("Could not marshal tree: %s", err), exitErrorDumpingTree)
-	}
-	_, err = tmpfile.Write(out)
-	if err != nil {
-		return nil, cli.NewExitError(fmt.Sprintf("Could not write output file: %s", err), exitCouldNotWriteOutputFile)
-	}
-	origHash, err := hashFile(tmpfile.Name())
-	if err != nil {
-		return nil, cli.NewExitError(fmt.Sprintf("Could not hash file: %s", err), exitCouldNotReadInputFile)
-	}
-	for {
-		err = runEditor(tmpfile.Name())
+	if c.String("set") != "" {
+		pathValuePair := strings.SplitAfterN(c.String("set"), "] ", 2)
+		if len(pathValuePair) < 2 {
+			return nil, cli.NewExitError("Invalid --set format", exitErrorInvalidSetFormat)
+		}
+		path := strings.TrimRight(pathValuePair[0], " ")
+		value := pathValuePair[1]
+		var valueTree interface{}
+		err := encodingjson.Unmarshal([]byte(value), &valueTree)
+		// Check if decoding it as json we find a single value
+		// and not a map or slice, in which case we can't marshal
+		// it to a sops.TreeBranch
+		kind := reflect.ValueOf(valueTree).Kind()
+		if kind == reflect.Map || kind == reflect.Slice {
+			var err error
+			valueTree, err = (&json.Store{}).Unmarshal([]byte(value))
+			if err != nil {
+				return nil, cli.NewExitError("Invalid --set value format", exitErrorInvalidSetFormat)
+			}
+		}
+		splitPath := strings.Split(path, "[")
+		parent, err := tree.Branch.Truncate(strings.Join(splitPath[0:len(splitPath)-1], ""))
 		if err != nil {
-			return nil, cli.NewExitError(fmt.Sprintf("Could not run editor: %s", err), exitNoEditorFound)
+			return nil, cli.NewExitError("Invalid --set path", exitErrorInvalidSetFormat)
 		}
-		newHash, err := hashFile(tmpfile.Name())
+		key, err := sops.TrimTreePathComponent(splitPath[len(splitPath)-1])
 		if err != nil {
-			return nil, cli.NewExitError(fmt.Sprintf("Could not hash file: %s", err), exitCouldNotReadInputFile)
+			return nil, cli.NewExitError("Invalid --set path format", exitErrorInvalidSetFormat)
 		}
-		if bytes.Equal(newHash, origHash) {
-			return nil, cli.NewExitError("File has not changed, exiting.", exitFileHasNotBeenModified)
+		branch := parent.(sops.TreeBranch)
+		for i, kv := range branch {
+			if kv.Key == key {
+				branch[i].Value = valueTree
+				break
+			}
 		}
```

Comment:
Would it be cleaner to move the above chunk of code to a function outside of `main()` ?